sklearn/metrics/pairwise.py

```diff
@@ -426,6 +426,57 @@ def pairwise_distances_argmin(X, Y, axis=1, metric="euclidean",
                                          metric_kwargs)[0]
 
 
+def haversine_distances(X, Y=None, ):
+    """
+    Compute haversine distance between samples in X and Y.
+
+    Haversine (Spherical) Distance
+
+    The Haversine distance is the angular distance between two points on
+    the surface of a sphere. The first distance of each point is assumed
+    to be the latitude, the second is the longitude, given in radians.
+    The dimension of the points must be 2.
+
+    .. math::
+       D(x, y) = 2\arcsin[\sqrt{\sin^2((x1 - y1) / 2)
+                                + cos(x1)cos(y1)sin^2((x2 - y2) / 2)}]
+
+    Parameters
+    ----------
+    X : array_like
+        with shape (n_samples_X, 2).
+
+    Y : array_like
+        with shape (n_samples_Y, 2).
+
+    Returns
+    -------
+    distance matrix : array
+        An array with shape (n_samples_X, n_samples_Y).
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.metrics.pairwise import haversine_distances
+    >>> X = np.array([[0, np.pi], [np.pi/2, 0]])
+    >>> Y = np.array([[-np.pi/2, 0], [-np.pi/2, -np.pi]])
+    >>> haversine_distances(X, Y) # doctest: +ELLIPSIS
+    array([[ 1.57...,  1.57...],
+           [ 3.14...,  3.14...]])
+    """
+
+    if Y is None:
+        Y = X
+    if (X.shape[1] != 2 or Y.shape[1] != 2):
+        raise ValueError("Haversine distance only valid in 2 dimensions")
```

Comment:
you should use [`check_paired_array`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/pairwise.py#L101) as done [here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/pairwise.py#L192).